python_scripts/concat_particles.py

```diff
@@ -0,0 +1,250 @@
+
+    # Compute the offset slicing for the 3D data
+    nx_local, ny_local, nz_local = source_file.attrs['dims_local']
+    x_start, y_start, z_start    = source_file.attrs['offset']
+    x_end, y_end, z_end          = x_start+nx_local, y_start+ny_local, z_start+nz_local
+
+    # Get the local number of particles
+    num_particles_local = source_file.attrs['n_particles_local'][0]
+
+    # write data from individual processor file to correct location in concatenated file
+    for dataset in list(destination_file.keys()):
+
+      if dataset == 'density':
+        destination_file[dataset][x_start:x_end,
+                                  y_start:y_end,
+                                  z_start:z_end] = source_file[dataset]
+      else:
+        start = particles_offset
+        end   = particles_offset + num_particles_local
+        destination_file[dataset][start:end] = source_file[dataset]
+
+    # Update the particles offset
+    particles_offset += num_particles_local
+
+    # Now that the copy is done we close the source file
+    source_file.close()
+
+  # Close destination file now that it is fully constructed
+  destination_file.close()
+# ==============================================================================
+
+# ==============================================================================
+def __get_num_particles(source_directory: pathlib.Path,
+                        num_processes: int,
+                        output_number: int) -> int:
+  """Get the total number of particles in the output. This function is heavily
+  I/O bound and might benefit from utilizing threads.
+
+  Parameters
+  ----------
+  source_directory : pathlib.Path
+      The directory of the unconcatenated files
+  num_processes : int
+      The number of processes
+  output_number : int
+      The output number to get data from
+
+  Returns
+  -------
+  int
+      The total number of particles in the output
+  """
+  # loop over files for a given output
+  num_particles = 0
+  for i in range(0, num_processes):
+    # open the input file for reading
+    with h5py.File(source_directory / f'{output_number}_particles.h5.{i}', 'r') as source_file:
+      num_particles += source_file.attrs['n_particles_local']
+
+  return num_particles
+# ==============================================================================
+
+# ==============================================================================
+def __setup_destination_file(source_directory: pathlib.Path,
+                             destination_file: h5py.File,
+                             output_number: int,
+                             num_particles: int,
+                             skip_fields: list,
+                             destination_dtype: np.dtype,
+                             compression_type: str,
+                             compression_options: str,
+                             chunking) -> h5py.File:
+  """_summary_
```

Comment:
Is this meant to have a short description of what this function does?

Reply:
Good catch. Fixed it. Thank you.